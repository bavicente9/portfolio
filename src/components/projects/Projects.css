.projects__container {
    padding: 0 var(--GeneralMarginWidth);
}

.projects__cardsContainer {
    display: flex;
    flex-direction: column;
}

.projects h1{
    color: var(--DarkBlue);
    text-transform: uppercase;
    text-align: center;
    font-size: 2rem;
    margin-bottom: 50px;
}

/*card*/
.projects__cardsContainer .projects__card:last-child{
    border-width: 1px 0 1px 0;
}
.projects__card {
    display: grid;
    grid-template-columns: 50% 50%;
    place-items: center;

    padding: 50px 0;
    border: 1px solid var(--Blue);
    border-width: 1px 0  0 0;
}

.projects__cardImage {

    aspect-ratio: 16/9;
    background: var(--GeneralBackGround);
    display: grid;
    place-items: center;
    border: 1px solid var(--Blue);
    border-radius: 25px;
}

.projects__cardImage img {
    height: 90%;
    width: 95%;
    -o-object-fit: cover;
       object-fit: cover;
    border-radius: 25px;
}

.projects__cardInfo {
    margin: 10px;
}

.projects__cardInfo h2 {
    font-size: 2rem;
    color: var(--Cyan);
    text-transform: capitalize;
}

.projects__cardInfoTechnologies {
    list-style: none;
    padding: 0;
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(100px, -webkit-max-content));
    grid-template-columns: repeat(auto-fit, minmax(100px, max-content));
    grid-template-rows: repeat(auto-fill, 1fr);
    gap: 10px;

}

.projects__cardInfoTechnologies li {
    display: inline;
    width: 100px;
    border: 1px solid var(--Cyan);
    border-width: 1px 0;

    font-family: 'Roboto', sans-serif;
    font-size: 1rem;
}

.projects__cardInfo p {
    font-family: 'Roboto', sans-serif;
    font-size: 1rem;
}

.projects__cardButtons {
    margin-top: 25px;
    display: flex;
    gap: 25px;
}

.projects__cardButtons .button {
    display: grid;
    place-items: center;
    text-decoration: none;
}

.projects__cardButtons .button--white {
    background: #fff;
    color: #000;
    border: 1px solid var(--Blue);
}

.projects__cardButtons .button--white:hover {
    background:var(--DarkGray);
}


/*Finish card*/


.projects__nav{
    margin: 25px 0;
    display: flex;
    gap: 15px;
    justify-content: center;
}

.projects__navButton{
    height: 30px;
    width: 30px;

    background: #fff;
    border: 1px solid var(--DarkCyan);
    border-radius: 5px;
    color: var(--DarkCyan);

    font-weight: 600;
    cursor: pointer;


}
.projects__navButton:hover{
    background:var(--DarkGray);
}
.projects__navButton--selected{
    background:var(--DarkCyan);
    color: #fff;
}
.projects__navButton--selected:hover{
    background:var(--DarkCyan);
}