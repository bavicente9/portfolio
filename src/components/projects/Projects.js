import { useEffect, useState } from 'react'
import './Projects.css'
import projectsInfo from './ProjectsInfo.json'



//return each card with its info according to the json
const generateCards = (page, infoJson) => {
    let cardsComponents = []
    for (let i = (page * 3 - 3); i < (page * 3); i++) {
        const item = infoJson[i]
        //check if there is not more items
        if (typeof item === 'undefined') break

        cardsComponents.push(
            <div key={i} className='projects__card'>
                <div className='projects__cardImage'>
                    <img alt='' src={item.image} />
                </div>
                <div className='projects__cardInfo'>
                    <h2>{item.name}</h2>
                    <ul className='projects__cardInfoTechnologies'>
                        {item.technologies.map((tech, index) => <li key={index}>{tech}</li>)}
                    </ul>
                    <p>{item.description}</p>
                    <nav className='projects__cardButtons' aria-label='project links'>
                        <a className='button' href={item.link_liveDemo} target="_blank">Live Demo</a>
                        <a className='button button--white' href={item.link_github} target="_blank">Github</a>
                    </nav>
                </div>
            </div>
        )
    }
    return (<>{cardsComponents}</>)

}

const generateNavButtons = (infoJson, handleChangePage) => {
    let buttonsComponents = []
    const nButtons = Math.ceil(infoJson.length / 3)

    //change the page and the button's state


    for (let i = 1; i <= nButtons; i++) {
        buttonsComponents.push(
            <button key={i} id={`projects__navButton${i}`} className='projects__navButton ' onClick={(e) => { handleChangePage(e.target.value) }} value={i}>{i}</button>
        )
    }

    return (<>{buttonsComponents}</>)
}


export const Projects = () => {
    const [page, setPage] = useState(1)
    const [btnSelected, setBtnSelected] = useState(1)

    //togle the class projects__navButton--selected 
    useEffect(() => {
        let prevButtonSelected = document.getElementsByClassName('projects__navButton--selected')
        let currentButtonSelected = document.getElementById(`projects__navButton${btnSelected}`)
        if (typeof prevButtonSelected[0] !== 'undefined') {
            prevButtonSelected[0].classList.remove('projects__navButton--selected')
        }
        currentButtonSelected.classList.add('projects__navButton--selected')
    }, [btnSelected])


    const handleChangePage = (value) => {
        setPage(value)
        setBtnSelected(value)
        document.getElementById('projects').scrollIntoView()
    }

    return (
        <section aria-label='projects' id='projects' className='projects'>
            <h1>Projects</h1>
            <div className='projects__container'>
                <div key={Math.random()} className='projects__cardsContainer'>
                    {generateCards(page, projectsInfo)}
                </div>
                <nav className='projects__nav' aria-label='Projects pages'>
                    {generateNavButtons(projectsInfo, handleChangePage)}
                </nav>
            </div>
        </section>
    )
}