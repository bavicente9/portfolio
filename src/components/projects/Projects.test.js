import { Projects } from "./Projects"
import { render, screen } from '@testing-library/react'
import userEvent from "@testing-library/user-event"

describe('Testing Projects', () => {
    const infoJson = JSON.parse(`[
        {
            "name": "project Name1",
            "image": "./images/maribelnails.png",
            "technologies": [
                "React",
                "JavaScript",
                "Css",
                "HTML"
            ],
            "description": "Here goes the project's description",
            "link_liveDemo": "https://github.com",
            "link_github": "https://github.com"
        },
        {
            "name": "project Name2",
            "image": "./images/maribelnails.png",
            "technologies": [
                "React",
                "JavaScript",
                "Css",
                "HTML"
            ],
            "description": "Here goes the project's description",
            "link_liveDemo": "https://github.com",
            "link_github": "https://github.com"
        },
        {
            "name": "project Name3",
            "image": "./images/maribelnails.png",
            "technologies": [
                "React",
                "JavaScript",
                "Css",
                "HTML"
            ],
            "description": "Here goes the project's description",
            "link_liveDemo": "https://github.com",
            "link_github": "https://github.com"
        }
        ,{
            "name": "project Name4",
            "image": "./images/maribelnails.png",
            "technologies": [
                "React",
                "JavaScript",
                "Css",
                "HTML"
            ],
            "description": "Here goes the project's description",
            "link_liveDemo": "https://github.com",
            "link_github": "https://github.com"
        }
     ]`)

    it('renders correctly', () => {
        const { asFragment } = render(<Projects infoJson={infoJson} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('render the correct nº of pages buttons', () => {
        render(<>
            <div id='skills'></div>
            <Projects infoJson={infoJson} />
        </>
        )
        const pagesBtns = document.getElementsByClassName('projects__navButton')
        expect(pagesBtns.length).toBe(2)
    })
    it('render the correct nº of cards', () => {
        render(<>
            <div id='skills'></div>
            <Projects infoJson={infoJson} />
        </>
        )
        const pageCards = document.getElementsByClassName('projects__card')
        expect(pageCards.length).toBe(3)
    })

    it('change rendered cards through the pages buttons', () => {
        //this line is cause jest in jsdom dont reconize the function: scrollIntoView
        window.HTMLElement.prototype.scrollIntoView = function() {};
        render(<>
            <Projects infoJson={infoJson} />
        </>
        )


        const pageBtn2 = document.getElementById('projects__navButton2')
        userEvent.click(pageBtn2)
        expect(screen.getByText('project Name4')).toBeInTheDocument()
    })

})