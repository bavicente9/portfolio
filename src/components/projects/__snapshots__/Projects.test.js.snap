// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Projects renders correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="projects"
    class="projects"
    id="projects"
  >
    <h1>
      Projects
    </h1>
    <div
      class="projects__container"
    >
      <div
        class="projects__cardsContainer"
      >
        <div
          class="projects__card"
        >
          <div
            class="projects__cardImage"
          >
            <img
              alt=""
              src="./images/maribelnails.png"
            />
          </div>
          <div
            class="projects__cardInfo"
          >
            <h2>
              project Name1
            </h2>
            <ul
              class="projects__cardInfoTechnologies"
            >
              <li>
                React
              </li>
              <li>
                JavaScript
              </li>
              <li>
                Css
              </li>
              <li>
                HTML
              </li>
            </ul>
            <p>
              Here goes the project's description
            </p>
            <nav
              aria-label="project links"
              class="projects__cardButtons"
            >
              <a
                class="button"
                href="https://github.com"
                target="_blank"
              >
                Live Demo
              </a>
              <a
                class="button button--white"
                href="https://github.com"
                target="_blank"
              >
                Github
              </a>
            </nav>
          </div>
        </div>
        <div
          class="projects__card"
        >
          <div
            class="projects__cardImage"
          >
            <img
              alt=""
              src="./images/maribelnails.png"
            />
          </div>
          <div
            class="projects__cardInfo"
          >
            <h2>
              project Name2
            </h2>
            <ul
              class="projects__cardInfoTechnologies"
            >
              <li>
                React
              </li>
              <li>
                JavaScript
              </li>
              <li>
                Css
              </li>
              <li>
                HTML
              </li>
            </ul>
            <p>
              Here goes the project's description
            </p>
            <nav
              aria-label="project links"
              class="projects__cardButtons"
            >
              <a
                class="button"
                href="https://github.com"
                target="_blank"
              >
                Live Demo
              </a>
              <a
                class="button button--white"
                href="https://github.com"
                target="_blank"
              >
                Github
              </a>
            </nav>
          </div>
        </div>
        <div
          class="projects__card"
        >
          <div
            class="projects__cardImage"
          >
            <img
              alt=""
              src="./images/maribelnails.png"
            />
          </div>
          <div
            class="projects__cardInfo"
          >
            <h2>
              project Name3
            </h2>
            <ul
              class="projects__cardInfoTechnologies"
            >
              <li>
                React
              </li>
              <li>
                JavaScript
              </li>
              <li>
                Css
              </li>
              <li>
                HTML
              </li>
            </ul>
            <p>
              Here goes the project's description
            </p>
            <nav
              aria-label="project links"
              class="projects__cardButtons"
            >
              <a
                class="button"
                href="https://github.com"
                target="_blank"
              >
                Live Demo
              </a>
              <a
                class="button button--white"
                href="https://github.com"
                target="_blank"
              >
                Github
              </a>
            </nav>
          </div>
        </div>
      </div>
      <nav
        aria-label="Projects pages"
        class="projects__nav"
      >
        <button
          class="projects__navButton projects__navButton--selected"
          id="projects__navButton1"
          value="1"
        >
          1
        </button>
        <button
          class="projects__navButton "
          id="projects__navButton2"
          value="2"
        >
          2
        </button>
      </nav>
    </div>
  </section>
</DocumentFragment>
`;
